feat: examples generator v1

czlowiek488 opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [Y] The commit message follows our guidelines: https://github.com/ngneat/falso/blob/master/CONTRIBUTING.md#commit
- [N] Tests for the changes have been added (for bug fixes / features)
- [Y] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[X] Documentation content changes
[X] Other... Please describe: CD related changes
```

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #268 

## What is the new behavior?

## Does this PR introduce a breaking change?

```
[ ] Yes
[X] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

I prepared a generator for examples since I don't want anyone to forget about updating those.
If you add `@automaticallyGeneratedExamples` in jsdoc and run generator, if the corresponding json file will be found it will put no more than 3 json file data elements after `@example` jsdoc property.
If there is no examples it will add new.
If there are already examples it will replace them.
Keep in mind that everything in comment after  `@automaticallyGeneratedExamples` will be replaced by examples.
Currently all files has well and consistent strucutre.
If something will be wrong with the files which are processed ( for example not consistent structure ) it will throw an error and stop generating next examples.

I hope we will be able to enjoy this feature soon :)

NetanelBasal commented:
What about functions such as dates, etc?

czlowiek488 commented:
If you wish to add examples for the functions without a json file I see 2 options
- add those manually
- add data file to them

On the other hand, instead of taking examples from data files we can generate examples using a rand function.
This solution would cover all the randomize functions however examples will be randomized and changed each release - and that's a reason why I decided to use static data, to give better control of what examples actually are.

NetanelBasal commented:
> This solution would cover all the randomize functions however examples will be randomized and changed each release 

I don't see any issue with this.

czlowiek488 commented:
I personally don't think it is a good idea.
I wouldn't like to give up control since we don't know what the future brings.

If you don't mind I would like to introduce a new json files instead.
Adding new files will make codebase more consistent.

However I am not gonna block this issue just because of what I belives, so if you wish to make it in a different way I will be fine with other solution.

NetanelBasal commented:
The disadvantage of json files is that we need to maintain them.

czlowiek488 commented:
Yeah sure but what's the real cost of maintance of those files?
I am not quiet familiar with changes which were introduced here in the past however, from my experience result of the functions that generates specific data does not change so often.

Now I think about hybrid option, where 1 attempt is to use a data from file, but 2 attempt is to generate examples from function itself.

What do you think?

NetanelBasal commented:
That's what I meant from the beginning. If we have a data file, use it; otherwise, produce it.

czlowiek488 commented:
I changed the way script is executed to give more flexability.
`./node_modules/.bin/ts-node ./scripts/example-generator.js ./packages/falso/src/lib/ @automaticallyGeneratedExamples`

NetanelBasal commented on package.json:
```diff
@@ -3,6 +3,7 @@
   "version": "0.0.0",
   "license": "MIT",
   "scripts": {
+    "prepublishOnly": "./node_modules/.bin/ts-node ./scripts/example-generator.js ./packages/falso/src/lib/ @automaticallyGeneratedExamples",
```
You can add it to the post-build.js script.

czlowiek488 replied:
Done

czlowiek488 commented:
Tell me if you see any more issues to resolve before adding generator indicator to other files.

czlowiek488 commented:
Is something wrong? I would like to finish it, do you have something in mind you would like to clarify before merging?

NetanelBasal commented:
Can you please fix the conflicts?

czlowiek488 commented:
I fixed conflicts and added indicator to every file with generator.
It looks pretty nice. However there is a bug.
Sometimes (73 times) when generator uses a function to generate an example `ts-node` throws an error about wrong types, even with `-T` flag builded js throws other errors.
I tried to fix it but I can't. I wish some would like to help here.
The result of script.
```
$ npm run generate-examples

> falso@0.0.0 generate-examples
> ./node_modules/.bin/ts-node ./scripts/example-generator.js ./packages/falso/src/lib/ @automaticallyGeneratedExamples

Generating examples...
packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/address.ts:55:7 - error TS2322: Type 'unknown' is not assignable to type 'string'.

55       city: randCity(),
         ~~~~

  packages/falso/src/lib/address.ts:15:3
    15   city: string;
         ~~~~
    The expected type comes from property 'city' which is declared here on type 'Address'
packages/falso/src/lib/address.ts:60:7 - error TS2322: Type 'unknown' is not assignable to type 'string | undefined'.
  Type 'unknown' is not assignable to type 'string'.

60       address.county = randCounty();
         ~~~~~~~~~~~~~~
packages/falso/src/lib/address.ts:64:7 - error TS2322: Type 'unknown' is not assignable to type 'string | undefined'.
  Type 'unknown' is not assignable to type 'string'.

64       address.country = randCountry();
         ~~~~~~~~~~~~~~~

packages/falso/src/lib/airport-code.ts:2:22 - error TS2732: Cannot find module './airport.json'. Consider using '--resolveJsonModule' to import module with '.json' extension.

2 import { data } from './airport.json';
                       ~~~~~~~~~~~~~~~~
packages/falso/src/lib/airport-code.ts:26:29 - error TS7031: Binding element 'code' implicitly has an 'any' type.

26     return rand(data.map(({ code }) => code));
                               ~~~~

packages/falso/src/lib/airport-name.ts:2:22 - error TS2732: Cannot find module './airport.json'. Consider using '--resolveJsonModule' to import module with '.json' extension.

2 import { data } from './airport.json';
                       ~~~~~~~~~~~~~~~~
packages/falso/src/lib/airport-name.ts:26:29 - error TS7031: Binding element 'name' implicitly has an 'any' type.

26     return rand(data.map(({ name }) => name));
                               ~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/bird.ts:2:22 - error TS2732: Cannot find module './bird.i18n.json'. Consider using '--resolveJsonModule' to import module with '.json' extension.

2 import { data } from './bird.i18n.json';
                       ~~~~~~~~~~~~~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/flight-details.ts:59:9 - error TS2322: Type '() => { passenger: string; airline: Airline; flightNumber: string; origin: unknown; destination: unknown; date: string; seat: string; flightLength: number; }' is not assignable to type '() => FlightDetails'.
  Call signature return types '{ passenger: string; airline: Airline; flightNumber: string; origin: unknown; destination: unknown; date: string; seat: string; flightLength: number; }' and 'FlightDetails' are incompatible.
    The types of 'origin' are incompatible between these types.
      Type 'unknown' is not assignable to type 'Airport'.

59   const factory: () => FlightDetails = () => {
           ~~~~~~~

packages/falso/src/lib/flight-number.ts:3:27 - error TS2732: Cannot find module './airline.json'. Consider using '--resolveJsonModule' to import module with '.json' extension.

3 import * as airlines from './airline.json';
                            ~~~~~~~~~~~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/hex.ts:2:10 - error TS2305: Module '"./random"' has no exported member 'random'.

2 import { random } from './random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/post.ts:37:7 - error TS2322: Type 'unknown' is not assignable to type 'string'.

37       title: randText({ charCount: 40 }),
         ~~~~~

  packages/falso/src/lib/post.ts:9:3
    9   title: string;
        ~~~~~
    The expected type comes from property 'title' which is declared here on type 'Post'
packages/falso/src/lib/post.ts:38:7 - error TS2322: Type 'unknown' is not assignable to type 'string'.

38       body: randText({ charCount: 500 }),
         ~~~~

  packages/falso/src/lib/post.ts:10:3
    10   body: string;
         ~~~~
    The expected type comes from property 'body' which is declared here on type 'Post'
packages/falso/src/lib/post.ts:39:7 - error TS2322: Type '{ id: string; text: unknown; user: User; }[]' is not assignable to type '{ id: string; text: string; user: User; }[]'.
  Type '{ id: string; text: unknown; user: User; }' is not assignable to type '{ id: string; text: string; user: User; }'.
    Types of property 'text' are incompatible.
      Type 'unknown' is not assignable to type 'string'.

39       comments: Array.from({ length: randNumber({ min: 1, max: 5 }) }, () => {
         ~~~~~~~~

  packages/falso/src/lib/post.ts:11:3
    11   comments: Array<{ id: string; text: string; user: User }>;
         ~~~~~~~~
    The expected type comes from property 'comments' which is declared here on type 'Post'

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
    at Object.getExamplesFromTs (/home/lukasz/workspace/packages/falsofork/falso/scripts/example-generator.js:46:31)
    at generateExamples (/home/lukasz/workspace/packages/falsofork/falso/scripts/example-generator.js:73:19)
    at Array.reduce (<anonymous>)
    at Object.<anonymous> (/home/lukasz/workspace/packages/falsofork/falso/scripts/example-generator.js:86:41)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at phase4 (/home/lukasz/workspace/packages/falsofork/falso/node_modules/ts-node/src/bin.ts:579:12)
packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/sequence.ts:2:10 - error TS2305: Module '"./random"' has no exported member 'random'.

2 import { random } from './random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/superhero-name.ts:3:22 - error TS2732: Cannot find module './superhero.json'. Consider using '--resolveJsonModule' to import module with '.json' extension.

3 import { data } from './superhero.json';
                       ~~~~~~~~~~~~~~~~~~
packages/falso/src/lib/superhero-name.ts:31:24 - error TS7031: Binding element 'company' implicitly has an 'any' type.

31       ? data.filter(({ company }) => company === options.company)
                          ~~~~~~~

packages/falso/src/lib/swift.ts:83:11 - error TS2322: Type 'unknown' is not assignable to type 'string'.

83     const countryCode: string = options?.countryCode ?? randCountryCode();
             ~~~~~~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/text.ts:46:12 - error TS2571: Object is of type 'unknown'.

46     while (text.length < charCount) {
              ~~~~
packages/falso/src/lib/text.ts:50:12 - error TS2571: Object is of type 'unknown'.

50     text = text.substring(0, charCount - 2);
              ~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/user.ts:52:37 - error TS2322: Type 'unknown' is not assignable to type 'string | undefined'.
  Type 'unknown' is not assignable to type 'string'.

52       email: randEmail({ firstName, lastName }),
                                       ~~~~~~~~
packages/falso/src/lib/user.ts:54:7 - error TS2322: Type 'unknown' is not assignable to type 'string'.

54       lastName,
         ~~~~~~~~

  packages/falso/src/lib/user.ts:14:3
    14   lastName: string;
         ~~~~~~~~
    The expected type comes from property 'lastName' which is declared here on type 'User'
packages/falso/src/lib/user.ts:57:43 - error TS2322: Type 'unknown' is not assignable to type 'string | undefined'.
  Type 'unknown' is not assignable to type 'string'.

57       username: randUserName({ firstName, lastName }),
                                             ~~~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

packages/falso/src/lib/core/core.ts:1:10 - error TS2305: Module '"../random"' has no exported member 'random'.

1 import { random } from '../random';
           ~~~~~~

<<-- Generator Errors -->>
--> 0 - '@automaticallyGeneratedExamples' is missing in jsdoc:
   []
--> 6 - corresponding ts file does not exists:
   [core, factories, gender-code, i18n, language-code, protocol-full]
--> 0 - 'json.data' field is missing:
   []
--> 5 - 'json.data' field value is not array:
   [code-snippet, first-name, food, last-name, sports]
--> 0 - 'json.data' field value is empty array:
   []
--> 0 - ts file must contain function which name starts with 'rand':
   []
--> 73 - fileReadError:
   [account, address, airport-code, airport-name, alpha-numeric, alpha, amount, arn, avatar, aws-request-id, between-date, bic, binary, bird, bitcoin-address, boolean, credit-card-cvv, credit-card, domain-name, email, flight-details, flight-number, float, font-size, full-address, full-name, future-date, git-commit-entry, git-commit-message, git-commit-sha, git-short-sha, hex, hexa-decimal, hsl, iban, img, ip, ipv6, json, latitude, line, longitude, mac, mask, nearby-gpscoordinate, number, octal, password, past-date, port, post, rand, random, recent-date, rgb, routing-number, seat-number, semver, sequence, slug, soon-date, sports-team, street-address, superhero-name, swift, text-range, text, todo, url, user-name, user, uuid, zip-code]

<<-- Generator Result -->>
--> generated: 117
--> files: 201
```
